### PR TITLE
Abstract local storage

### PR DESCRIPTION
**Pull request by nicolas-brousse**
# Description:

This PR intent to abstract current local storage, and make easier to add custom storage (like S3 storage).

I didn't finish test yet, but I may work on it if you're interested on this feature.

Related to #107, #143, #159, #242

______________

# Tasks:

-  Describe the problem / feature
-  Write tests

I will abide by the [code of conduct](https://github.com/gemstash/gemstash/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

File: .rubocop-bundler.yml [hunk 1]
```diff
@@ -27,7 +27,7 @@ Layout/EndAlignment:
 Layout/AccessModifierIndentation:
   EnforcedStyle: outdent
 
-Layout/AlignParameters:
+Layout/ParameterAlignment:
```
Comment (bronzdoc) [inline .rubocop-bundler.yml:30]: Can you move rubocop changes to a separate PR?

File: .rubocop-bundler.yml [hunk 2]
```diff
@@ -103,7 +103,7 @@ Naming/HeredocDelimiterNaming:
 # This should be enabled, but test broke if so, see:
 # spec/support/simple_server.rb:149:20
 # spec/support/slow_simple_server.rb:15:20
-Naming/UncommunicativeMethodParamName:
```
Comment (bronzdoc) [inline .rubocop-bundler.yml:106]: Same here

File: gemstash.gemspec [hunk 3]
```diff
@@ -18,9 +18,9 @@ you push your own private gems as well."
   spec.homepage      = "https://github.com/rubygems/gemstash"
   spec.license       = "MIT"
 
-  spec.files         = `git ls-files -z`.split("\x0").select {|f|
+  spec.files         = `git ls-files -z`.split("\x0").select do |f|
```
Comment (bronzdoc) [inline gemstash.gemspec:21]: You can submit a PR for this, is not related to the goal of this PR

File: spec/data/gems/speaker/exe/speaker [hunk 4]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env ruby
 
-# frozen_string_literal: tru
+# frozen_string_literal: true
```
Comment (bronzdoc) [inline spec/data/gems/speaker/exe/speaker:3]: 👍

File: spec/gemstash/http_client_spec.rb [hunk 5]
```diff
@@ -82,7 +82,7 @@
       context "with a client that specifies a user agent" do
         let(:http_client) do
           Gemstash::HTTPClient.new(faraday_client,
-            user_agent: "my-agent 6.6.6")
+                                   user_agent: "my-agent 6.6.6")
```
Comment (bronzdoc) [inline spec/gemstash/http_client_spec.rb:85]: You can improve the style of piece of code if you are modifying it for the purpose of the PR otherwise this is better suited for another PR 🙂

File: spec/gemstash/storage_spec.rb [hunk 6]
```diff
@@ -15,13 +15,6 @@
     expect(Gemstash::Storage.new(@folder)).not_to be_nil
   end
 
-  it "builds the path if it does not exists" do
```
Comment (bronzdoc) [inline spec/gemstash/storage_spec.rb:18]: Why we don't need this specs?

File: spec/gemstash/upstream_spec.rb [hunk 7]
```diff
@@ -59,7 +59,7 @@
 
   it "supports getting user agent" do
     expect(Gemstash::Upstream.new("https://rubygems.org/",
-      user_agent: "my_user_agent").user_agent).to eq("my_user_agent")
+                                  user_agent: "my_user_agent").user_agent).to eq("my_user_agent")
```
Comment (bronzdoc) [inline spec/gemstash/upstream_spec.rb:62]: Same style comment here

File: lib/gemstash/upstream.rb [hunk 8]
```diff
@@ -15,7 +15,7 @@ class Upstream
     def initialize(upstream, user_agent: nil)
       @uri = URI(CGI.unescape(upstream.to_s))
       @user_agent = user_agent
-      raise "URL '#{@uri}' is not valid!" unless @uri.to_s =~ URI::DEFAULT_PARSER.make_regexp
+      raise "URL '#{@uri}' is not valid!" unless @uri.to_s&.match?(URI::DEFAULT_PARSER.make_regexp)
```
Comment (bronzdoc) [inline lib/gemstash/upstream.rb:18]: Why this is needed?

File: rakelib/changelog.rake [hunk 9]
```diff
@@ -37,7 +37,7 @@ class Changelog
     return if master_update?
 
     contents = File.read(changelog_file)
-    return unless contents =~ /^## master \(unreleased\)$/
+    return unless /^## master \(unreleased\)$/.match?(contents)
```
Comment (bronzdoc) [inline rakelib/changelog.rake:40]: Why is this needed?

File: lib/gemstash/gem_fetcher.rb [hunk 10]
```diff
@@ -8,7 +8,7 @@ module Gemstash
   class GemFetcher
     def initialize(http_client)
       @http_client = http_client
-      @valid_headers = Set.new(["etag", "content-type", "content-length", "last-modified"])
+      @valid_headers = Set.new(%w[etag content-type content-length last-modified])
```
Comment (bronzdoc) [inline lib/gemstash/gem_fetcher.rb:11]: Why is this needed?

**Comment by nicolas-brousse** [comment]
@bronzdoc thanks for your review, I'll try to have a look on it this week.

**Comment by mobilutz** [comment]
@nicolas-brousse Did you by any chance look into this?

Would love to have a possibility to run gemstash on heroku ;)
Or course, a S3 adapter is still missing, but this would be the big first step!

**Comment by nicolas-brousse** [comment]
Oopsy. I totally forgot about it.

Lot of work in this period. I'll try to found some time in the following days.

Thanks for the ping @mobilutz 😅

File: lib/gemstash/upstream.rb (hunk 8 above)
Comment (nicolas-brousse) [reply lib/gemstash/upstream.rb:18]: I'll revert it. Its change is related to rubocop warnings.

File: rakelib/changelog.rake (hunk 9 above)
Comment (nicolas-brousse) [reply rakelib/changelog.rake:40]: I'll revert it. Its change is related to rubocop warnings.

**Comment by bronzdoc** [comment]
Hi everyone 👋 

Our GSOC student James Nguyen is working on the S3 backend for Gemstash this year. As part of that feature he also worked on the abstraction of local storage. 

We will split those two things in two different PR's to avoid confusion please follow https://github.com/rubygems/gemstash/pull/270

Reviews and feedback is appreciated, thanks

File: lib/gemstash/env.rb [hunk 11]
```diff
@@ -177,5 +177,16 @@ def cache_client
...
     end
+
+    def storage_service
+      @storage_service ||= begin
+        case config[:storage_adapter]
+        when "local"
+          Storage::LocalService.new
+        else
+          raise "Invalid storage service: '#{config[:storage_adapter]}'"
+        end
```
Comment (mobilutz) [inline lib/gemstash/env.rb:188]: How about making this a bit more dynamic, maybe something like this:
```
config[:storage_adapter] = 'Storage::LocalService' # so a class name
```
```suggestion
        Object.const_get(config[:storage_adapter]).new
```

Of course, there needs to be some more checks and some form of `raise`, but it would make sense to make this less hard coded.

Comment (nicolas-brousse) [reply lib/gemstash/env.rb:188]: It's okay for me, I can apply this change.

Comment (nicolas-brousse) [reply lib/gemstash/env.rb]: I did the change :)

File: lib/gemstash/gem_fetcher.rb (hunk 10 above)
Comment (nicolas-brousse) [reply lib/gemstash/gem_fetcher.rb]: I reverted the change since it was related to a rubocop warning.

File: spec/gemstash/storage_spec.rb (hunk 6 above)
Comment (nicolas-brousse) [reply spec/gemstash/storage_spec.rb:18]: I re-add the test and change a bit the test itself since I moved the folder creation.

File: spec/gemstash/http_client_spec.rb (hunk 5 above)
Comment (nicolas-brousse) [reply spec/gemstash/http_client_spec.rb]: I'll create an other PR for this :)

**Comment by nicolas-brousse** [comment]
@bronzdoc I have some wip in local too. I didn't push it yet since it's based on this work. But I may share this work later if you want.

**Comment by nicolas-brousse** [comment]
I did some changes.

Let me know it there is more to change.

If you plan to merge this PR before #270, then I will wait for this one merged before. Then I may take time to help you on review #270 🙂

File: lib/gemstash/storage.rb [hunk 12]
```diff
@@ -27,23 +27,22 @@ def initialize(folder, version)
...
+      Resource.new(self.class.storage_service, @folder, id)
     end
 
     # Fetch a nested entry from this instance in the storage engine.
     #
     # @param child [String] the name of the nested entry to load
     # @return [Gemstash::Storage] a new storage instance for the +child+
     def for(child)
-      Storage.new(File.join(@folder, child), root: false)
+      self.class.new(File.join(@folder, child), root: false)
```
Comment (bronzdoc) [inline lib/gemstash/storage.rb:45]: 🤔  I don't understand this change, why we need it? is the same as calling `Storage.new` ?

File: lib/gemstash/storage.rb [hunk 13]
```diff
@@ -133,9 +128,9 @@ def initialize(folder, name)
     # @return [Boolean] true if the indicated content exists
     def exist?(key = nil)
       if key
-        File.exist?(properties_filename) && File.exist?(content_filename(key))
+        @service.exist?(properties_filename) && @service.exist?(content_filename(key))
```
Comment (bronzdoc) [inline lib/gemstash/storage.rb:131]: Why a service should be responsible of checking if a file exist?

File: lib/gemstash/storage.rb [hunk 14]
```diff
@@ -250,13 +245,13 @@ def delete(key)
...
       begin
-        File.delete(content_filename(key))
+        @service.delete(content_filename(key))
       rescue StandardError => e
         log_error "Failed to delete stored content at #{content_filename(key)}", e, level: :warn
       end
 
       begin
-        File.delete(properties_filename) unless content?
+        @service.delete(properties_filename) unless content?
```
Comment (bronzdoc) [inline lib/gemstash/storage.rb:254]: Why a service should be responsible of deleting files?

File: lib/gemstash/storage/local_service.rb [hunk 15]
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+require "fileutils"
+
+module Gemstash
+  class Storage
+    class LocalService < BaseService # :nodoc:
```
Comment (bronzdoc) [inline lib/gemstash/storage/local_service.rb:7]: I think we don't need exits?, read and delete I feel we just coupling the a `Resource` with a service by forcing to using these service methods. What you think?

File: lib/gemstash/upstream.rb [hunk 16]
```diff
@@ -15,7 +15,7 @@ class Upstream
     def initialize(upstream, user_agent: nil)
       @uri = URI(CGI.unescape(upstream.to_s))
       @user_agent = user_agent
-      raise "URL '#{@uri}' is not valid!" unless @uri.to_s =~ URI::DEFAULT_PARSER.make_regexp
+      raise "URL '#{@uri}' is not valid!" unless @uri.to_s  =~ URI::DEFAULT_PARSER.make_regexp
```
Comment (bronzdoc) [inline lib/gemstash/upstream.rb:18]: This just add an extra space? before `=~` ?

**Comment by duyquangnguyenhac** [comment]
Hi everyone, just wanna say that I'm looking forward to seeing this PR merged so that I can use it in my PR #270!

**Comment by nicolas-brousse** [comment]
Hey! Sorry I've got two really busy weeks. I'm in holidays for two weeks now, I'll try to found a bit of time to continue this work.

**Comment by bronzdoc** [comment]
Im going to close this since the abstraction of the local storage was made in https://github.com/rubygems/gemstash/pull/270
